File: app/src/main/java/com/drdisagree/iconify/ui/adapters/ClockPreviewAdapter.java
package com.drdisagree.iconify.ui.adapters;

import static com.drdisagree.iconify.common.Preferences.LSCLOCK_SWITCH;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.drdisagree.iconify.R;
import com.drdisagree.iconify.config.RPrefs;
import com.drdisagree.iconify.ui.models.ClockModel;
import com.drdisagree.iconify.ui.utils.ViewBindingHelpers;
import com.drdisagree.iconify.utils.WallpaperUtil;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.imageview.ShapeableImageView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Objects;

@SuppressWarnings("deprecation")
public class ClockPreviewAdapter extends RecyclerView.Adapter<ClockPreviewAdapter.ViewHolder> {

    private final Context context;
    private final ArrayList<ClockModel> itemList;
    private final String prefSwitch;
    private final String prefStyle;
    private static Bitmap wallpaperBitmap;
    private LinearLayoutManager linearLayoutManager;

    public ClockPreviewAdapter(Context context, ArrayList<ClockModel> itemList, String prefSwitch, String prefStyle) {
        this.context = context;
        this.itemList = itemList;
        this.prefSwitch = prefSwitch;
        this.prefStyle = prefStyle;

        new WallpaperLoaderTask(this).execute();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.view_clock_preview, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        ClockModel model = itemList.get(position);

        holder.bind(model, position);

        if (wallpaperBitmap != null) {
            ViewBindingHelpers.setBitmap(holder.wallpaperView, wallpaperBitmap);
        }
    }

    @Override
    public int getItemCount() {
        return itemList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        private final LinearLayout container;
        private final TextView title;
        private final LinearLayout clockContainer;
        private final ImageView checkIcon;
        private final MaterialButton button;
        private final ShapeableImageView wallpaperView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            container = itemView.findViewById(R.id.clock_preview_child);
            title = itemView.findViewById(R.id.clock_title);
            clockContainer = itemView.findViewById(R.id.clock_view_container);
            checkIcon = itemView.findViewById(R.id.icon_selected);
            button = itemView.findViewById(R.id.btn_select_style);
            wallpaperView = itemView.findViewById(R.id.wallpaper_view);
        }

        public void bind(ClockModel model, int position) {
            title.setText(model.getTitle());

            button.setOnClickListener(v -> {
                int index = position + (Objects.equals(prefSwitch, LSCLOCK_SWITCH) ? 0 : 1);
                RPrefs.putInt(prefStyle, index);
                refreshLayout(this);
            });

            int adapterPosition = getAdapterPosition();
            if (RPrefs.getInt(prefStyle, 0) != adapterPosition) {
                checkIcon.setVisibility(View.GONE);
                button.setEnabled(true);
            } else {
                checkIcon.setVisibility(View.VISIBLE);
                button.setEnabled(false);
            }

            clockContainer.removeAllViews();
            ViewStub viewStub = new ViewStub(context);
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            );
            params.gravity = Gravity.CENTER_VERTICAL;
            viewStub.setLayoutParams(params);
            viewStub.setLayoutResource(model.getLayout());
            clockContainer.addView(viewStub);

            if (viewStub.getParent() != null) {
                viewStub.inflate();
            }

            boolean isSelected = adapterPosition == getBindingAdapterPosition();
            button.setEnabled(!isSelected);
            checkIcon.setVisibility(isSelected ? View.VISIBLE : View.GONE);
        }
    }

    private void refreshLayout(ViewHolder holder) {
        int firstVisible = linearLayoutManager.findFirstVisibleItemPosition() - 1;
        int lastVisible = linearLayoutManager.findLastVisibleItemPosition() + 1;

        for (int i = firstVisible; i <= lastVisible; i++) {
            View view = linearLayoutManager.findViewByPosition(i);
            if (view == null) continue;

            LinearLayout child = view.findViewById(R.id.clock_preview_child);
            if (child == null) continue;

            boolean isSelected = view == holder.container;
            child.findViewById(R.id.btn_select_style).setEnabled(!isSelected);
            child.findViewById(R.id.icon_selected).setVisibility(!isSelected ?
                    View.GONE :
                    View.VISIBLE
            );
        }
    }

    @Override
    public void onViewAttachedToWindow(@NonNull ViewHolder holder) {
        super.onViewAttachedToWindow(holder);

        if (RPrefs.getInt(prefStyle, 0) != holder.getBindingAdapterPosition()) {
            holder.checkIcon.setVisibility(View.GONE);
            holder.button.setEnabled(true);
        } else {
            holder.checkIcon.setVisibility(View.VISIBLE);
            holder.button.setEnabled(false);
        }
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);

        linearLayoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
    }

    private static class WallpaperLoaderTask extends AsyncTask<Void, Void, Bitmap> {
        private final WeakReference<ClockPreviewAdapter> adapterRef;

        WallpaperLoaderTask(ClockPreviewAdapter adapter) {
            adapterRef = new WeakReference<>(adapter);
        }

        @Override
        protected Bitmap doInBackground(Void... voids) {
            Context context = adapterRef.get().context;
            if (context == null) return null;
            return WallpaperUtil.getCompressedWallpaper(context, 80);
        }

        @SuppressLint("NotifyDataSetChanged")
        @Override
        protected void onPostExecute(Bitmap bitmap) {
            super.onPostExecute(bitmap);
            ClockPreviewAdapter adapter = adapterRef.get();
            if (adapter != null && bitmap != null) {
                wallpaperBitmap = bitmap;
                adapter.notifyDataSetChanged();
            }
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/ui/adapters/ClockPreviewAdapter.kt
package com.drdisagree.iconify.ui.adapters

import android.annotation.SuppressLint
import android.app.WallpaperManager
import android.content.Context
import android.graphics.Bitmap
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.drdisagree.iconify.R
import com.drdisagree.iconify.common.Preferences.LSCLOCK_SWITCH
import com.drdisagree.iconify.config.RPrefs
import com.drdisagree.iconify.ui.models.ClockModel
import com.drdisagree.iconify.ui.utils.ViewBindingHelpers.setBitmapWithAnimation
import com.drdisagree.iconify.utils.WallpaperUtils
import com.google.android.material.button.MaterialButton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@Suppress("deprecation")
class ClockPreviewAdapter(
    private val context: Context,
    private val itemList: ArrayList<ClockModel>,
    prefSwitch: String?,
    prefStyle: String,
    private val mOnStyleSelected: OnStyleSelected? = null
) : RecyclerView.Adapter<ClockPreviewAdapter.ViewHolder>() {

    private val prefStyle: String
    private var linearLayoutManager: LinearLayoutManager? = null

    init {
        Companion.prefSwitch = prefSwitch
        this.prefStyle = prefStyle
        loadWallpaper(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(
            if (prefSwitch == LSCLOCK_SWITCH) {
                R.layout.view_clock_preview_lockscreen
            } else {
                R.layout.view_clock_preview_header
            },
            parent,
            false
        )
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val model = itemList[position]
        holder.bind(model, position)
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val container: LinearLayout = itemView.findViewById(R.id.clock_preview_child)
        private val title: TextView = itemView.findViewById(R.id.clock_title)
        private val clockContainer: LinearLayout = itemView.findViewById(R.id.clock_view_container)
        val checkIcon: ImageView = itemView.findViewById(R.id.icon_selected)
        val button: MaterialButton = itemView.findViewById(R.id.btn_select_style)
        private val wallpaperView: ImageView = itemView.findViewById(R.id.wallpaper_view)

        fun bind(model: ClockModel, position: Int) {
            title.text = model.title

            button.setOnClickListener {
                RPrefs.putInt(prefStyle, position)
                mOnStyleSelected?.onStyleSelected(position)
                refreshLayout(this)
            }

            clockContainer.removeAllViews()

            val viewStub = ViewStub(context)
            val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )

            params.gravity = Gravity.CENTER_VERTICAL
            viewStub.setLayoutParams(params)
            viewStub.layoutResource = model.layout

            clockContainer.addView(viewStub)

            if (viewStub.parent != null) {
                viewStub.inflate()
            }

            val isSelected = adapterPosition == getBindingAdapterPosition()
            button.setEnabled(!isSelected)
            checkIcon.setVisibility(if (isSelected) View.VISIBLE else View.GONE)

            if (wallpaperBitmap != null) {
                setBitmapWithAnimation(wallpaperView, wallpaperBitmap)
            }
        }
    }

    private fun refreshLayout(holder: ViewHolder) {
        val firstVisible = linearLayoutManager!!.findFirstVisibleItemPosition() - 1
        val lastVisible = linearLayoutManager!!.findLastVisibleItemPosition() + 1

        for (i in firstVisible..lastVisible) {
            val view = linearLayoutManager!!.findViewByPosition(i) ?: continue
            val child = view.findViewById<LinearLayout>(R.id.clock_preview_child) ?: continue
            val isSelected = view === holder.container

            child.findViewById<View>(R.id.btn_select_style).setEnabled(!isSelected)
            child.findViewById<View>(R.id.icon_selected).visibility =
                if (!isSelected) View.GONE else View.VISIBLE
        }
    }

    override fun onViewAttachedToWindow(holder: ViewHolder) {
        super.onViewAttachedToWindow(holder)

        if (RPrefs.getInt(prefStyle, 0) != holder.getBindingAdapterPosition()) {
            holder.checkIcon.setVisibility(View.GONE)
            holder.button.setEnabled(true)
        } else {
            holder.checkIcon.setVisibility(View.VISIBLE)
            holder.button.setEnabled(false)
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)

        linearLayoutManager = recyclerView.layoutManager as LinearLayoutManager?
    }

    @SuppressLint("NotifyDataSetChanged")
    fun loadWallpaper(adapter: ClockPreviewAdapter) {
        CoroutineScope(Dispatchers.Main).launch {
            val bitmap = withContext(Dispatchers.IO) {
                WallpaperUtils.getCompressedWallpaper(
                    adapter.context,
                    80,
                    if (prefSwitch == LSCLOCK_SWITCH) {
                        WallpaperManager.FLAG_LOCK
                    } else {
                        WallpaperManager.FLAG_SYSTEM
                    }
                )
            }

            if (bitmap != null && prefSwitch == LSCLOCK_SWITCH) {
                wallpaperBitmap = bitmap
                adapter.notifyDataSetChanged()
            }
        }
    }

    interface OnStyleSelected {
        /**
         * Interface for style selection
         * @param position The position of the selected style,
         * in our case is the num of the layout
         */
        fun onStyleSelected(position: Int)
    }

    companion object {
        private var prefSwitch: String? = null
        private var wallpaperBitmap: Bitmap? = null
    }
}
